compiler/src/iree/compiler/Codegen/Common/GPU/GPUDecomposeVectorExtOps.cpp

```diff
@@ -0,0 +1,127 @@
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler {
+
+using namespace IREE::VectorExt;
+
+namespace {
+
+struct DecomposeThreadIds : public OpRewritePattern<ThreadIdsOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ThreadIdsOp threadIdsOp,
+                                PatternRewriter &rewriter) const override {
+    Value tid = threadIdsOp.getTid();
+    NestedLayoutAttr layout =
+        dyn_cast<NestedLayoutAttr>(threadIdsOp.getLayout());
+    if (!layout) {
+      return rewriter.notifyMatchFailure(threadIdsOp,
+                                         "NestedLayoutAttr expected.");
+    }
+
+    SmallVector<Value> virtualTids;
+    int64_t rank = layout.getRank();
+    // The delinearized thread IDs are returned from outer most to inner most,
+    // i.e. before applying the layout described dimensions ordering.
+    SmallVector<Value> threadIds = layout.computeThreadIds(tid, rewriter);
+
+    SmallVector<Value> filteredThreadIds;
+    for (auto [id, active] : llvm::zip(llvm::drop_begin(threadIds, rank),
+                                       layout.getThreadActiveIds())) {
+      if (active)
+        filteredThreadIds.push_back(id);
+    }
+
+    // Subgroup and thread (lane) indices normalized to the order in which
+    // they are used by each dimension.
+    virtualTids = llvm::to_vector(
+        llvm::map_range(invertPermutationVector(layout.getThreadOrder()),
+                        [&](int64_t i) { return filteredThreadIds[i]; }));
+
+    rewriter.replaceOp(threadIdsOp, virtualTids);
+    return success();
+  }
+};
+
+struct DecomposeSubgroupIds : public OpRewritePattern<SubgroupIdsOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  DecomposeSubgroupIds(MLIRContext *ctx, int64_t subgroupSize,
+                       PatternBenefit benefit = 1)
+      : OpRewritePattern(ctx, benefit), subgroupSize(subgroupSize) {}
+
+  LogicalResult matchAndRewrite(SubgroupIdsOp subgroupIdsOp,
+                                PatternRewriter &rewriter) const override {
+    Value tid = subgroupIdsOp.getTid();
+    NestedLayoutAttr layout =
+        dyn_cast<NestedLayoutAttr>(subgroupIdsOp.getLayout());
+    if (!layout) {
+      return rewriter.notifyMatchFailure(subgroupIdsOp,
+                                         "NestedLayoutAttr expected.");
+    }
+
+    SmallVector<Value> virtualTids;
+    // The delinearized thread IDs are returned from outer most to inner most,
+    // i.e. before applying the layout described dimensions ordering.
+    SmallVector<Value> threadIds = layout.computeThreadIds(tid, rewriter);
+
+    SmallVector<Value> filteredSubgroupIds;
+    for (auto [id, active] :
+         llvm::zip(threadIds, layout.getSubgroupActiveIds())) {
+      if (active)
+        filteredSubgroupIds.push_back(id);
```

Comment:
I was trying to split this patch up: https://github.com/iree-org/iree/pull/17284 , which does this distribution independent of each other, which is why I made them separate operations.

In general, it makes much more sense to me for these to be decoupled. The computeIds function we have on layouts should also decouple these ideally.

Reply:
I don't see that patch doing thread/subgroup distribution independently, but we should also probably rename `subgroup` distribution because that's not how it's actually being used right now. The layout setting logic for transfer read ops treats "thread" and "subgroup" ids homogeneously so they can't happen in different stages.

Reply:
Let's discuss offline, getting a bit hard to explain my reasoning on text about this :P